### PR TITLE
Add Gnuplot builder

### PR DESCRIPTION
**Pull request by VarLad**
This again is incomplete! 
@giordano Your kind help is yet again needed my friend!

### CODE REVIEW

**Comment by clad26** [comment]
There's an executable called gnuplot-qt, generated in libexec folder
Should I label it as ExecutableProduct or something like LibraryExecutableProduct? Would BinaryBuilder still recognize it?

**Comment by giordano** [comment]
https://juliapackaging.github.io/BinaryBuilderBase.jl/dev/#BinaryBuilderBase.ExecutableProduct

> By default, the library is searched in the `bindir`, but you can specify a different directory within the prefix with the `dir_path` argument.

**Comment by clad26** [comment]
> https://juliapackaging.github.io/BinaryBuilderBase.jl/dev/#BinaryBuilderBase.ExecutableProduct
> 
> > By default, the library is searched in the `bindir`, but you can specify a different directory within the prefix with the `dir_path` argument.

So...     ExecutableProduct(dir_path="libexec/gnuplot_qt", :gnuplot_qt)
???

**Comment by giordano** [comment]
No, `ExecutableProduct("gnuplot_qt", :gnuplot_qt, dir_path="libexec")`.

**Comment by clad26** [comment]
```
ERROR: LoadError: MethodError: no method matching ExecutableProduct(::String, ::Symbol; dir_path="libexec")
Closest candidates are:
  ExecutableProduct(::AbstractString, ::Symbol, ::Any...) at /depot/packages/BinaryBuilderBase/xqk4x/src/Products.jl:331 got unsupported keyword argument "dir_path"
  ExecutableProduct(::Vector{String}, ::Symbol) at /depot/packages/BinaryBuilderBase/xqk4x/src/Products.jl:321 got unsupported keyword argument "dir_path"
  ExecutableProduct(::Vector{String}, ::Symbol, ::Union{Nothing, AbstractString}) at /depot/packages/BinaryBuilderBase/xqk4x/src/Products.jl:321 got unsupported keyword argument "dir_path"
Stacktrace:
 [1] top-level scope
   @ /agent/_work/1/s/G/Gnuplot/build_tarballs.jl:25
in expression starting at /agent/_work/1/s/G/Gnuplot/build_tarballs.jl:25
##[error]Bash exited with code '1'.
```

**Comment by giordano** [comment]
Ugh, yeah, my fault, `dir_path` isn't a keyword argument: `ExecutableProduct("gnuplot_qt", :gnuplot_qt, "libexec")`

**Comment by clad26** [comment]
Should I use FileProduct()?

**Comment by clad26** [comment]
I'm mainly seeing 3 errors....
The libiconv one
``` 
/opt/x86_64-linux-gnu/bin/../lib/gcc/x86_64-linux-gnu/4.8.5/../../../../x86_64-linux-gnu/bin/ld: encoding.o: undefined reference to symbol 'libiconv'
/opt/x86_64-linux-gnu/bin/../x86_64-linux-gnu/sys-root/usr/local/lib/libiconv.so.2: error adding symbols: DSO missing from command line

```

The Qt one
```
[17:28:49] make[4]: /workspace/destdir/bin/lrelease: No such file or directory
[17:28:49] make[4]: *** [Makefile:1498: qtgnuplot_fr.qm] Error 127
[17:28:49] make[4]: *** Waiting for unfinished jobs....

```

and the cairo one

```
[17:28:49] ../term/write_png_image.c:81:10: fatal error: cairo-pdf.h: No such file or directory
[17:28:49]  #include "cairo-pdf.h"
[17:28:49]           ^~~~~~~~~~~~~
[17:28:49] compilation terminated.
```

**Comment by clad26** [comment]
> [17:28:49] ../term/write_png_image.c:81:10: fatal error: cairo-pdf.h: No such file or directory
[17:28:49]  #include "cairo-pdf.h"
[17:28:49]           ^~~~~~~~~~~~~
[17:28:49] compilation terminated.

This is weird...
```
sandbox:${WORKSPACE}/destdir # find ${prefix} -name cairo-pdf.h
/workspace/destdir/include/cairo/cairo-pdf.h

```

**Comment by clad26** [comment]
@giordano 
I fixed it with
```
make CFLAGS=-I${prefix}/include/cairo
```
But there are more such folders within include
How to list all the folders in include?

File: G/Gnuplot/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,41 @@
+
+using BinaryBuilder, Pkg
+
+name = "gnuplot"
+version = v"5.4.1"
+
+# Collection of sources required to complete build
+sources = [
+    ArchiveSource("https://excellmedia.dl.sourceforge.net/project/gnuplot/gnuplot/5.4.1/gnuplot-5.4.1.tar.gz", "6b690485567eaeb938c26936e5e0681cf70c856d273cc2c45fabf64d8bc6590e"),
```
Comment (fingolfin) [inline G/Gnuplot/build_tarballs.jl:9]: Use $version, move checksum to second line?

Comment (clad26) [reply G/Gnuplot/build_tarballs.jl]: Okay, done!

**Comment by giordano** [comment]
I have the feeling gnuplot can't be easily cross-compiled

**Comment by clad26** [comment]
> I have the feeling gnuplot can't be easily cross-compiled

It might serve as a reference
Screenshot from 2021-01-03 00-29-01

**Comment by clad26** [comment]
@giordano Are you talking about Windows or MacOS?

**Comment by giordano** [comment]
I don't see anything useful in that screenshot.

The fact that it's trying to run a command, `lrelease`, worries me: it's a binary program, likely built for the target platform, it isn't very useful to us (in addition to the fact it isn't even in the `Qt_jll` artifact).

**Comment by clad26** [comment]
https://github.com/barche/QML.jl/issues/104

**Comment by clad26** [comment]
@giordano 
I'm opening a new PR to make a binary for this https://github.com/qt/qttools
It uses qt as a dependency, which I think we already have

**Comment by clad26** [comment]
https://github.com/JuliaPackaging/Yggdrasil/pull/2340

This is a dependency to build the above package!

**Comment by clad26** [comment]
1. I'll look at qt terminal in the longer term, due to build problems
2. As of current, I've implemented X11, but will remove that in favour of wxt terminal, as its available on multiple platforms, unlike x11.

**Comment by clad26** [comment]
@giordano Is there a way I can use this environment, without the PR, in my own forked repo?

**Comment by fingolfin** [comment]
@clad26 You can use the `Project.toml` in the `.ci` dir (see https://github.com/JuliaPackaging/Yggdrasil/pull/2325#issuecomment-754966578); you also will need a specific development version of Julia for now, see e.g. https://github.com/JuliaPackaging/Yggdrasil/pull/2320#issuecomment-754960800 (hopefully when Julia 1.6 comes out, this will become simpler again)

File: G/Gnuplot/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    BuildDependency(PackageSpec(name="Xorg_xorgproto_jll", uuid = "c4d99508-4286-5418-9131-c86396af500b")),
+    Dependency(PackageSpec(name="Libcerf_jll", uuid="af83a40a-c4c4-57a0-81df-2309fbd279e3")),
+    Dependency(PackageSpec(name="LibGD_jll", uuid="16339573-6216-525a-b38f-30b6f6b71b5f")),
+    #Dependency(PackageSpec(name="X11_jll", uuid="546b0b6d-9ca3-5ba2-8705-1bc1841d8479")),
+    #After WxWidgets has compiled
+    #Dependency(PackageSpec(name="WxWidgets_jll", uuid="")),
+    Dependency(PackageSpec(name="Cairo_jll", uuid="83423d85-b0ee-5818-9007-b63ccbeb887a")),
+    Dependency(PackageSpec(name="Pango_jll", uuid="36c8627f-9965-5494-a995-c6b170f724f3")),
```
Comment (fingolfin) [inline G/Gnuplot/build_tarballs.jl:44]: Perhaps also add `Readline_jll`?

File: G/Gnuplot/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+sources = [
+    ArchiveSource("https://excellmedia.dl.sourceforge.net/project/gnuplot/gnuplot/$(version)/gnuplot-$(version).tar.gz",
+                  "6b690485567eaeb938c26936e5e0681cf70c856d273cc2c45fabf64d8bc6590e"),
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/gnuplot-*/
+# Make uic happy to use `libc.so` ¯\_(ツ)_/¯
+apk add g++ linux-headers samurai gnutls readline musl
```
Comment (fingolfin) [inline G/Gnuplot/build_tarballs.jl:17]: That's very weird, why is that needed? OK, we currently need `apk add samurai` as a workaround. But the rest seems suspicious to me.

**Comment by fingolfin** [comment]
The following might be useful:
- https://forum.archive.openwrt.org/viewtopic.php?id=69417
- https://sourceforge.net/p/gnuplot/support-requests/62/

File: G/Gnuplot/build_tarballs.jl (hunk 3 above)
Comment (giordano) [reply G/Gnuplot/build_tarballs.jl:17]: I don't know, there is something very wrong with `uic` in Qt: https://github.com/JuliaPackaging/Yggdrasil/blob/67f10c6e61ef7b0581f6a9a3d83dea9dbc260692/Q/Qt/build_tarballs.jl#L39.  I don't really understand why it requires `libc.so` instead of `libc.so.6` like 99.9999...% of executables.  At the moment this line isn't needed at all since we aren't using Qt

Comment (giordano) [reply G/Gnuplot/build_tarballs.jl:17]: See also https://github.com/JuliaPackaging/Yggdrasil/commit/e7967da3858f44d846d49efb66395d7228dd371d#r44397545

Comment (clad26) [reply G/Gnuplot/build_tarballs.jl:17]: gnuplot is actually available for alpine.
I was thinking about, and adding the same deps.
Doesn't work out.
I'll try adding Readline_jll

**Comment by clad26** [comment]
> The following might be useful:
> 
>     * https://forum.archive.openwrt.org/viewtopic.php?id=69417
> 
>     * https://sourceforge.net/p/gnuplot/support-requests/62/

THANKS!
Will try that....

**Comment by clad26** [comment]
``` https://sourceforge.net/p/gnuplot/support-requests/62/ ```
Doing this produces binaries, which give libiconv.so.2 not found error...

**Comment by barche** [comment]
With Qt5Tools_jll built and soon available, it should be possible to build Qt support again.

**Comment by VarLad** [comment]
@giordano x86_64 musl error seems to be 
`undefined reference to random.h`
While random.h is a part of linux-headers you added....
It should be in /usr/include/linux/random.h

Regarding Windows....
I don't think I can compile for Windows. 
Good thing is, for 64-bit Windows, there's gp541-win64-mingw.7z provided by gnuplot. Is there a way I can just download it and extract its contents in `$WORKSPACE/destdir`?
For 32-bit Windows, I'll try disabling qt and enabling the windows terminal.

**Comment by VarLad** [comment]
MSys version is still kinda limited

https://sourceforge.net/projects/gnuplot/files/gnuplot/5.4.1/gp541-win64-mingw.7z/download is complete and self-contained for Windows 64 bit. Its provided by gnuplot and is the latest version

https://sourceforge.net/projects/gnuplot/files/gnuplot/5.2.7/gp527-win32-mingw.7z/download is the same. It's not the latest version, but perfectly usable.

**Comment by VarLad** [comment]
Note that MSys2 version doesn't build for Qt, but for WxWidgets. And the binary for that isn't complete yet.....

**Comment by VarLad** [comment]
@giordano Also note that the Windows binary can't seem to find Lua or LibGD

**Comment by VarLad** [comment]
> Good thing is, for 64-bit Windows, there's gp541-win64-mingw.7z provided by gnuplot. Is there a way I can just download it and extract its contents in $WORKSPACE/destdir?

Is this really not possible? Its easier to rely on pre-built binaries in the case of Windows, especially just because its a "complete build" (meaning it has almost everything, all terminals etc in 30mb).

**Comment by VarLad** [comment]
It's almost there. Some problems
1. Think about what to do with Windows. (Recommended: Use the given binaries)
2. The `Products` section is incomplete as of yet. I think I'll need the wizard to complete that! Since different platforms will dish out different products
3. If pre-built binaries is not possible for Windows, then we'll have to compile WxWidgets. We can skip musl in wxwidgets right now. Again, the `Products` page of WxWidgets is incomplete. Also, we'll have to see why Lua and LibGD aren't being found by `./configure` found for Windows.

**Comment by giordano** [comment]
> Is this really not possible?

It is possible and there plenty of examples already (Git, HDF5, etc...), but I'm trying to understand why building from source doesn't work.  When building for Windows they reference functions implemented in a file which is never compiled to object file or used as source for other object files.  It doesn't make sense.

**Comment by VarLad** [comment]
@giordano If this experience taught me anything, that would be more appreciation for the mere existence of Rust and Zig...

**Comment by barche** [comment]
Pretty sure we're close to getting this to work, it seems the patch adds winmain.c to the list of sources but that doesn't get picked up for some reason.

**Comment by barche** [comment]
Even though this compiles, the resulting Windows JLL is very broken. Test version at: https://github.com/barche/Gnuplot_jll.jl

* First, it complains about Lua not existing.
* After removing Lua from the JLL, Pango errors with the following trace:

```
ERROR: LoadError: LoadError: InitError: could not load library "C:\Users\user\.julia\artifacts\d44308a9abb790d928ac029b7f1660d167ea67df\bin\libpango-1.0-0.dll"
%1 is not a valid Win32 application.
Stacktrace:
  [1] dlopen(s::String, flags::UInt32; throw_error::Bool)
    @ Base.Libc.Libdl .\libdl.jl:114
  [2] dlopen(s::String, flags::UInt32)
    @ Base.Libc.Libdl .\libdl.jl:114
  [3] macro expansion
    @ ~\.julia\packages\JLLWrappers\WnWcZ\src\products\library_generators.jl:63 [inlined]
  [4] __init__()
    @ Pango_jll ~\.julia\packages\Pango_jll\XQBSg\src\wrappers\x86_64-w64-mingw32.jl:16
  [5] _include_from_serialized(path::String, depmods::Vector{Any})
    @ Base .\loading.jl:674
  [6] _require_from_serialized(path::String)
    @ Base .\loading.jl:727
  [7] _require(pkg::Base.PkgId)
    @ Base .\loading.jl:1031
  [8] require(uuidkey::Base.PkgId)
    @ Base .\loading.jl:914
  [9] require(into::Module, mod::Symbol)
    @ Base .\loading.jl:901
 [10] include(mod::Module, _path::String)
    @ Base .\Base.jl:386
 [11] top-level scope
    @ ~\.julia\packages\JLLWrappers\WnWcZ\src\toplevel_generators.jl:170
 [12] include
    @ .\Base.jl:386 [inlined]
 [13] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt64}}, source::Nothing)
    @ Base .\loading.jl:1213
 [14] top-level scope
    @ none:1
 [15] eval
    @ .\boot.jl:360 [inlined]
 [16] eval(x::Expr)
    @ Base.MainInclude .\client.jl:446
 [17] top-level scope
    @ none:1
during initialization of module Pango_jll
```

File: G/Gnuplot/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    BuildDependency(PackageSpec(name="Xorg_xorgproto_jll", uuid = "c4d99508-4286-5418-9131-c86396af500b")),
+    #Dependency(PackageSpec(name="Libcerf_jll", uuid="af83a40a-c4c4-57a0-81df-2309fbd279e3")),
+    Dependency(PackageSpec(name="LibGD_jll", uuid="16339573-6216-525a-b38f-30b6f6b71b5f")),
+    Dependency(PackageSpec(name="Lua", uuid="a4086b1d-a96a-5d6b-8e4f-2030e6f25ba6")),
+    Dependency(PackageSpec(name="Qt5Tools_jll", uuid="a9c6e4b1-b2fb-56d5-96a9-25f276f13840")),
+    Dependency(PackageSpec(name="Qt5Svg_jll", uuid="3af4ccab-a251-578e-a514-ea85a0ba79ee")),
+    Dependency(PackageSpec(name="Qt5Base_jll", uuid="ea2cea3b-5b76-57ae-a6ef-0a8af62496e1")),
```
Comment (giordano) [inline G/Gnuplot/build_tarballs.jl:56]: Do we need all of these, especially at runtime?  The [archlinux package](https://archlinux.org/packages/extra/x86_64/gnuplot/) doesn't use `qt5-base`, and `qt5-tools` is a build dependency only

Comment (barche) [reply G/Gnuplot/build_tarballs.jl:56]: Correct, tools should be a build-dep, and Qt5Svg already pulls in Qt5Base.

**Comment by barche** [comment]
OK, I'm stuck now. The resulting Windows executable doesn't work. Sequence used for testing:

```julia
add https://github.com/barche/Gnuplot_jll.jl.git
using Gnuplot_jll
ENV["PATH"] = "$(Gnuplot_jll.PATH[]);$(ENV["PATH"])"
run(`gnuplot`)
```
Result: 
```
ERROR: failed process: Process(`gnuplot`, ProcessExited(3221225781)) [3221225781]
```

**Comment by VarLad** [comment]
That _might_ be because the `Products` part is incomplete

**Comment by barche** [comment]
> @barche That _might_ be because the `Products` part is incomplete
> 
> I need the Wizard for that. Uhhh... if you've got the internet data to run the wizard, can you please fill it up？

No, that is not the problem, `gnuplot` is the only executable created by the build, and it works when I compile for Mac. This must be something like a missing library or something similar, but I don't know how to make Windows give me a clearer error message. I don't think we need to run the wizard on this to improve it.

**Comment by VarLad** [comment]
@barche Yeah, sorry, just remembered that `Products` is just there for insurance

**Comment by barche** [comment]
> @barche Yeah, sorry, just remembered that `Products` is just there for insurance

No need to apologise, we're building binaries, so nobody know everything ;)

The products are used to make libraries and other files available from Julia, so the `ExecutableProduct` here gives you a `gnuplot` function in Julia that can be used to call the executable.

**Comment by VarLad** [comment]
@barche By the way, I was checking the Pipelines..... 
Windows doesn't build any terminals..... except the default ones.... 
We'll have to think for pango, cairo, and libgd later again(or I doubt it'll run in Jupyter/Pluto and some functions won't work)

Also, are you sure the qt build is okay? I see a lot of warnings!

**Comment by barche** [comment]
I did a test after making the new Qt packages that on macOS I could run the Gnuplot Qt terminal built with this. Good point about the other terminal types, this will probably require some further iteration to get it to work. Right now the Windows test binary only has the Qt terminal.

**Comment by VarLad** [comment]
Again, that patch is from MSys2.... and they've disabled qt in their build, instead they've enabled wxwidgets

**Comment by barche** [comment]
The warnings about Qt are deprecation warnings, so these should be harmless. Also, the archive contains the Qt terminal, so I'm sure it gets built. And gnuplot loads the terminals on-demand, the command line doesn't even appear, so something else is wrong.

**Comment by ViralBShah** [comment]
Is this good to merge?

**Comment by VarLad** [comment]
@barche What do you think, are the windows binaries stable/working?

**Comment by VarLad** [comment]
> MSys version is still kinda limited
> 
> https://sourceforge.net/projects/gnuplot/files/gnuplot/5.4.1/gp541-win64-mingw.7z/download is complete and self-contained for Windows 64 bit. Its provided by gnuplot and is the latest version
> 
> https://sourceforge.net/projects/gnuplot/files/gnuplot/5.2.7/gp527-win32-mingw.7z/download is the same. It's not the latest version, but perfectly usable.

@barche we can still do this if it still doesn't work

**Comment by ViralBShah** [comment]
It could be possible to get it merged, and then have the packages that use GNUPlot to allow the usage of these binaries, while being able to use system binaries through an environment variable that ignores the JLLs - and that way we can get them tested and update the binaries as necessary going forward.

**Comment by VarLad** [comment]
> > https://sourceforge.net/projects/gnuplot/files/gnuplot/5.4.1/gp541-win64-mingw.7z/download is complete and self-contained for Windows 64 bit. Its provided by gnuplot and is the latest version
> > https://sourceforge.net/projects/gnuplot/files/gnuplot/5.2.7/gp527-win32-mingw.7z/download is the same. It's not the latest version, but perfectly usable.

@giordano Lets just do this for completeness for now, to merge, and we can keep trying to compile for Windows later on?

**Comment by barche** [comment]
I haven't continued working on this, the Windows binaries built here don't work and I don't know why and I have no idea on how to diagnose the problem. It has nothing to do with Qt, since the program fails even before showing the gnuplot command prompt.

**Comment by giordano** [comment]
What about other platforms?  Is it only Windows which is not working?

**Comment by VarLad** [comment]
I think so, yeah
@barche mentioned that he tested for MAC

**Comment by barche** [comment]
I did some further debugging, and as it turns out my procedure for launching gnuplot was incorrect. The binaries actually do work. Here is an example test sequence:


In Pkg mode: 
```
add https://github.com/barche/Gnuplot_jll.jl.git
```

In the Julia prompt:
```julia 
using Gnuplot_jll
ENV["GNUPLOT_DRIVER_DIR"] = joinpath(Gnuplot_jll.artifact_dir, "libexec", "gnuplot", "5.4")
gnuplot() do x
       run(`$(Gnuplot_jll.gnuplot_path) --persist test.gnuplot`)
end
```

The file `test.gnuplot` is a gnuplot command file in the current working dir, e.g:

```gnuplot
# set terminal pngcairo  transparent enhanced font "arial,10" fontscale 1.0 size 600, 400 
# set output 'simple.1.png'
set key fixed left top vertical Right noreverse enhanced autotitle box lt black linewidth 1.000 dashtype solid
set samples 50, 50
set title "Simple Plots" 
set title  font ",20" textcolor lt -1 norotate
set xrange [ * : * ] noreverse writeback
set x2range [ * : * ] noreverse writeback
set yrange [ * : * ] noreverse writeback
set y2range [ * : * ] noreverse writeback
set zrange [ * : * ] noreverse writeback
set cbrange [ * : * ] noreverse writeback
set rrange [ * : * ] noreverse writeback
NO_ANIMATION = 1
plot [-10:10] sin(x),atan(x),cos(atan(x))
```

What is still strange is that launching gnuplot without arguments does not produce a gnuplot prompt, but I suppose we won't be using the interactive gnuplot terminal from Julia anyway.

**Comment by VarLad** [comment]
@barche Dang! It actually works for Windows that way
But there's still one tiny problem.... no lua/tikz terminal

It can't seem to recognize lua/tikz for some reason

**Comment by giordano** [comment]
> In the Julia prompt:
> ```julia 
> using Gnuplot_jll
> ENV["GNUPLOT_DRIVER_DIR"] = joinpath(Gnuplot_jll.artifact_dir, "libexec", "gnuplot", "5.4")
> gnuplot() do x
>        run(`$(Gnuplot_jll.gnuplot_path) --persist test.gnuplot`)
> end
> ```

The syntax to run executables provided by JLL packages is [documented](https://docs.binarybuilder.org/stable/jll/#ExecutableProduct):
```julia
run(`$(gnuplot()) --persist test.gnuplot`) # Julia v1.6+
gnuplot() do x # Julia v1.5-
    run(`$(x) --persist test.gnuplot`)
end
```
It's also good to _not_ set environment variables globally but limit their scope with `withenv` (Julia v1.5-) or `addenv` (Julia v1.6+)

**Comment by barche** [comment]
> @barche I think how Gaston.jl does it is, it opens gnuplot and sends commands to it via pipeline.....
> Do you think it'd work if the prompt doesn't work?

I don't know, someone will have to try :)

> Other than that, the 3d plots in qt terminal would lose their interactivity (rotation and zooming)

Interactivity works if you run with:
```julia
run(`$(gnuplot()) test.gnuplot -`)
```

I think the best way forward is to merge this and then try to use it for a bit, to see what needs to be added (Lua support, for example).

**Comment by ViralBShah** [comment]
Is it possible to revive this and get it merged - possibly with a solution for windows if things have improved?

**Comment by barche** [comment]
> Is it possible to revive this and get it merged - possibly with a solution for windows if things have improved?

As far as I can tell this was already OK. The problem is that it's not clear how the binary is used exactly, so I can't test much locally. Maybe a rebuild needs to be triggered given the time that has passed.

**Comment by ViralBShah** [comment]
I suppose we should update to the latest - 5.4.2.

**Comment by ViralBShah** [comment]
My update to use the latest patch from mingw and latest stable version isn't working on windows. Seems like some simple linking thing on windows.

**Comment by barche** [comment]
Took it one step further by forcing the winmain to be included, but this fails with:

```
win/wd2d.cpp:53:12: fatal error: prntvpt.h: No such file or directory
```

**Comment by ViralBShah** [comment]
Nice! Is it good to merge?

File: G/Gnuplot/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                  "66f679115dd30559e110498fc94d926949d4d370b4999a042e724b8e910ee478"),
+    DirectorySource("./bundled"),
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/gnuplot-*/
+
+# Don't try this at home, it's evil
+ln -s /opt/${host_target}/${host_target}/sys-root/usr/lib/libc.so /usr/lib/libc.so
```
Comment (giordano) [inline G/Gnuplot/build_tarballs.jl:19]: I don't think we still need it, right?

**Comment by barche** [comment]
I don't think its working, unfortunately. Trying to run it gives:

```
test-gnuplot\bin> .\gnuplot.exe
Program 'gnuplot.exe' failed to run: The specified executable is not a valid application for this OS platform.At line:1
 char:1
+ .\gnuplot.exe
+ ~~~~~~~~~~~~~.
At line:1 char:1
+ .\gnuplot.exe
+ ~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (:) [], ApplicationFailedException
    + FullyQualifiedErrorId : NativeCommandFailed
```

This is just starting the generated executable after unpacking the tarball generated by BB, not sure if that misses any needed setup.

**Comment by giordano** [comment]
https://docs.binarybuilder.org/stable/jll/#ExecutableProduct

**Comment by barche** [comment]
Failing with:

```
julia> run(`$(gnuplot()) --persist test.gnuplot`)
ERROR: IOError: could not spawn setenv(`'C:\Users\bartj\.julia\artifacts\66b1adb6061c27e0a5d34c0205c66359efd41fac\bin\gnuplot.exe' --persist test.gnuplot`,["PATH=C:\\Users\\bartj\\.julia\\artifacts\\231a53dac6fa0e5293789aac33fff0f9c0cdd033\\bin;C:\\Users\\bartj\\.julia\\artifacts\\4a2cc4b6aef321d12f704be234ed46372e67d1a2\\bin;C:\\Users\\bartj\\AppData\\Local\\Programs\\Julia-1.8.3\\bin;C:\\Users\\bartj\\.julia\\artifacts\\72037f1415327f12277eed0521c861517630a0ba\\bin;C:\\Users\\bartj\\.julia\\artifacts\\2e9a29903fddf79371cd5f82327f70ada670a234\\bin;C:\\Users\\bartj\\.julia\\artifacts\\3a34dec1472dcb90db22005558267071d9a8b913\\bin;C:\\Users\\bartj\\.julia\\artifacts\\01b969d96551292e78f0acfa204f120b09f4e7b4\\bin;C:\\Users\\bartj\\.julia\\artifacts\\5cbe7d14731ff4f280e1cdaa99b8d3b804f85ac0\\bin;C:\\Users\\bartj\\.julia\\artifacts\\e79b7522cd23369ab4425a478a2066d67e0df557\\bin;C:\\Users\\bartj\\.julia\\artifacts\\320fd25151d281a979541cea41d5df48e0e119f0\\bin;C:\\Users\\bartj\\.julia\\artifacts\\4c321ea83282f557b9ca0389a21d1227f428c35c\\bin;C:\\Users\\bartj\\.julia\\artifacts\\3764237d6451c548336209f3dc7ffa814198327d\\bin;C:\\Users\\bartj\\.julia\\artifacts\\8867311b9387088945ca854fdbd5ed302ce2e0e1\\bin;C:\\Users\\bartj\\.julia\\artifacts\\41eb25ea50f277f03154d52354602687fb1abc84\\bin;C:\\Users\\bartj\\.julia\\artifacts\\be75bce183282b09d5afa393777e1c4d09e36f6c\\bin;C:\\Users\\bartj\\.julia\\artifacts\\e5872943f1078a6397a4f4d0f760db959ec50630\\bin;C:\\Users\\bartj\\.julia\\artifacts\\027f99dde7e18a8ed89e38f995de8336a7130a3c\\bin;C:\\Users\\bartj\\.julia\\artifacts\\c38e0f0854753b42d208bf7f62aafa3100ba3880\\bin;C:\\Users\\bartj\\.julia\\artifacts\\5e6cdd29fca31b08026313af46c1db0c724ebbb2\\bin;C:\\Users\\bartj\\.julia\\artifacts\\2d50da999117a9f26877c370b0c06ce7f119be67\\bin;C:\\Users\\bartj\\.julia\\artifacts\\48a7da60b90abc7b20623102f7822d036bdeea75\\bin;C:\\Users\\bartj\\.julia\\artifacts\\c34a53538bc6c87c8b376c3f6d0d3248830003bc\\bin;C:\\Users\\bartj\\.julia\\artifacts\\4c30542eefae67acb7d51b8938627d01905b98ce\\bin;C:\\Users\\bartj\\.julia\\artifacts\\6a5456c156251deff41ae40ac8dfbb3d8c808ffe\\bin;C:\\Users\\bartj\\.julia\\artifacts\\3628616b5a5a8fd8628079ae4fb0d8948fdb37bc\\bin;C:\\Users\\bartj\\.julia\\artifacts\\266782898a47bef5cd0441562f9a5e3dae9fc9a2\\bin;C:\\Users\\bartj\\AppData\\Local\\Programs\\Julia-1.8.3\\bin\\..\\lib\\julia;C:\\Users\\bartj\\AppData\\Local\\Programs\\Julia-1.8.3\\bin\\..\\lib;C:\\Users\\bartj\\AppData\\Local\\Programs\\Julia-1.8.3\\bin ..., "PROCESSOR_ARCHITECTURE=AMD64"]): unknown error (UNKNOWN)
Stacktrace:
 [1] _spawn_primitive(file::String, cmd::Cmd, stdio::Vector{Union{RawFD, Base.Libc.WindowsRawSocket, IO}})
   @ Base .\process.jl:128
 [2] #725
   @ .\process.jl:139 [inlined]
 [3] setup_stdios(f::Base.var"#725#726"{Cmd}, stdios::Vector{Union{RawFD, Base.Libc.WindowsRawSocket, IO}})
   @ Base .\process.jl:223
 [4] _spawn
   @ .\process.jl:138 [inlined]
 [5] run(::Cmd; wait::Bool)
   @ Base .\process.jl:479
 [6] run(::Cmd)
   @ Base .\process.jl:477
 [7] top-level scope
   @ REPL[5]:1

```

The strange thing is that when I double-click the exe, I get a popup claiming "this app can't run on your PC". When I do the same on the old v5.4.1 exe, I get the expected "lib...dll not found" errors.

**Comment by barche** [comment]
p.s. If anyone wants to check the binaries, they are here: https://github.com/barche/Gnuplot_jll.jl/releases/tag/Gnuplot-v5.4.5%2B0

I'm still debugging this too.

**Comment by mkitti** [comment]
Remind me to take a look on Windows.

**Comment by barche** [comment]
So I found out that there is another simpler executable being built, which also doesn't run. And the trigger is just adding the `-L` flag:

Works:
```
cc -g -O2  -DHAVE_CONFIG_H -I.. -o bf_test_good.exe bf_test.c 
```

Fails:
```
cc -g -O2  -L/workspace/destdir/lib -DHAVE_CONFIG_H -I.. -o bf_test_bad.exe bf_test.c
```

The working version runs and produces no output, with the bad version I get:

```
Program 'bf_test_bad.exe' failed to run: The specified executable is not a valid application for this OS platform.At li
ne:1 char:1
+ .\bf_test_bad.exe
+ ~~~~~~~~~~~~~~~~~.
At line:1 char:1
+ .\bf_test_bad.exe
+ ~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (:) [], ApplicationFailedException
    + FullyQualifiedErrorId : NativeCommandFailed
```

Dependency Walker shows both exes only link to KERNEL32.DLL and MSVCRT.DLL. The built files differ in size.

**Comment by barche** [comment]
It seems one of the dependencies puts the following libraries in `/workspace/destdir/lib`:

```
libgcc.a  libgcc_s.a  libmsvcrt.a
```

Removing these builds a working executable, but I'm going to investigate which dep does this.

**Comment by barche** [comment]
Binaries are working now. The above-mentioned libraries were from `CompilerSupportLibraries_jll`, so I think the real bug is in that package. Current workaround is to simply remove `libgcc.a  libgcc_s.a  libmsvcrt.a`. For me this is ready to merge now.

File: G/Gnuplot/build_tarballs.jl (hunk 5 above)
Comment (barche) [reply G/Gnuplot/build_tarballs.jl]: Right, removed.

**Comment by giordano** [comment]
> The above-mentioned libraries were from `CompilerSupportLibraries_jll`, so I think the real bug is in that package. Current workaround is to simply remove `libgcc.a libgcc_s.a libmsvcrt.a`.

It'd be good to know _what_ is the problem with those libraries since we're going to use them in https://github.com/JuliaLang/julia/pull/47184.  CC: @vchuravy

**Comment by ViralBShah** [comment]
@barche If you have merge privileges, please merge. Otherwise, let me know and I can.

**Comment by barche** [comment]
> @barche If you have merge privileges, please merge. Otherwise, let me know and I can.

I don't, so please, go ahead!
